proposals/NNNN-no-implicit-copy-bindings.md

```diff
@@ -0,0 +1,487 @@
+we want the operation to be concise and consistently spelled.
+
+## Future directions
+
+### Expanding `@noImplicitCopy`'s applicability
+
+The `@noImplicitCopy` attribute can be useful in more contexts than we
+initially propose to support, including:
+
+#### `var` bindings and `inout` parameters
+
+Mutable variables and `inout` parameters could also be marked as `@noImplicitCopy`.
+This could be taken to mean that the current value bound to the variable cannot
+be implicit copied, but it can be taken or replaced with a new value, which
+would then also be restricted from being copied out of the variable.
+
+#### Scopes
+
+Since a motivating use case for this feature is performance control in
+critical sections or hot parts of code, it could be useful to apply a blanket
+"no implicit copies" constraint to a scope, instead of having to individually
+annotate every binding used in the scope.
+
+#### Types
+
+A type may need to be copyable, in order to satisfy protocol requirements or
+just because it does have a copy operation that makes sense, but it may be
+undesirable to copy it in most situations. Therefore, `@noImplicitCopy` may
+make sense as a type-level annotation as a way to suppress implicit copying
+any values that are statically of the type, while still allowing the type
+to be copyable in a way compatible with other copyable types. (An alternative
+might be to have the type be move-only but conform to some `ManuallyCopyable`
+protocol, but such types would not be able to interact generically with
+existing Swift APIs using built-in copyability..)
+
+#### Properties
+
+It would be useful to tag properties of a struct or class, or cases of an enum,
+as `@noImplicitCopy`. This would have the consequence of preventing accesses
+to the property from introducing implicit copies.
+
+### Variations of language features that borrow instead of consume
+
+It is obviously not ideal for basic operations like binding new variables,
+pattern-matching values, and iterating sequences to always consume their
+value. We plan to extend the language with the ability to perform these
+operations on borrowed and/or `inout` exclusively accessed values:
+
+- `borrow` and `inout` bindings can reference a value without copying,
+  respectively borrowing a read-only view of the value or exclusively accessing
+  it for mutation:
+
+    ```
+    @noImplicitCopy let x = getValue()
+    do {
+      borrow y = x // not a copy, but a borrow
+
+      use(x) // we can use both x and y
+      use(y)
+
+      consume(x) // ERROR: but we can't consume x while y is borrowing it
+      use(y)
+    }
+    ```
+
+- Pattern matching operations should also exist to project out of `Optional`
+  values and other enums without consuming the enum:
+
+    ```
+    @noImplicitCopy let x: Optional = getValue()
+
+    if borrow y = x {}
+
+    // We can still use x because we only borrowed its payload
+
+    switch x {
+    case .some(borrow y):
```

Comment:
Was shadowing 'y' intentional here?

Reply:
There isn't any shadowing, really,  since `if borrow y` would be out of scope by the time we reach the switch. But I can still change the name to be clear.